src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -2333,192 +1455,108 @@ class HandshakeForDeflation : public HandshakeClosure {
-  deflate_common_idle_monitors(self, false /* !is_global */, target);
-}
 
-// Deflate global or per-thread idle ObjectMonitors.
-//
-void ObjectSynchronizer::deflate_common_idle_monitors(Thread* self,
-                                                      bool is_global,
-                                                      JavaThread* target) {
-  int deflated_count = 0;
-  ObjectMonitor* free_head_p = NULL;  // Local SLL of scavenged ObjectMonitors
-  ObjectMonitor* free_tail_p = NULL;
-  ObjectMonitor* saved_mid_in_use_p = NULL;
-  elapsedTimer timer;
-
-  if (log_is_enabled(Info, monitorinflation)) {
-    timer.start();
-  }
-
-  if (is_global) {
-    OM_PERFDATA_OP(MonExtant, set_value(Atomic::load(&om_list_globals._in_use_count)));
-  } else {
-    OM_PERFDATA_OP(MonExtant, inc(Atomic::load(&target->om_in_use_count)));
-  }
-
-  do {
-    int local_deflated_count;
-    if (is_global) {
-      local_deflated_count =
-          deflate_monitor_list(self, &om_list_globals._in_use_list,
-                               &om_list_globals._in_use_count,
-                               &free_head_p, &free_tail_p,
-                               &saved_mid_in_use_p);
-    } else {
-      local_deflated_count =
-          deflate_monitor_list(self, &target->om_in_use_list,
-                               &target->om_in_use_count, &free_head_p,
-                               &free_tail_p, &saved_mid_in_use_p);
+      if (ls != NULL) {
+        ls->print_cr("after handshaking: in_use_list stats: ceiling=" SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
+                     in_use_list_ceiling(), _in_use_list.count(), _in_use_list.max());
+        timer.start();
+      }
     }
-    deflated_count += local_deflated_count;
-
-    if (free_head_p != NULL) {
-      // Move the deflated ObjectMonitors to the global free list.
-      guarantee(free_tail_p != NULL && local_deflated_count > 0, "free_tail_p=" INTPTR_FORMAT ", local_deflated_count=%d", p2i(free_tail_p), local_deflated_count);
-      // Note: The target thread can be doing an om_alloc() that
-      // is trying to prepend an ObjectMonitor on its in-use list
-      // at the same time that we have deflated the current in-use
-      // list head and put it on the local free list. prepend_to_common()
-      // will detect the race and retry which avoids list corruption,
-      // but the next field in free_tail_p can flicker to marked
-      // and then unmarked while prepend_to_common() is sorting it
-      // all out.
-#ifdef ASSERT
-      ObjectMonitor* l_next_om = unmarked_next(free_tail_p);
-      assert(l_next_om == NULL, "must be NULL: _next_om=" INTPTR_FORMAT, p2i(l_next_om));
-#endif
 
-      prepend_list_to_global_wait_list(free_head_p, free_tail_p, local_deflated_count);
+    // After the handshake, safely free the ObjectMonitors that were
+    // deflated in this cycle.
+    size_t deleted_count = 0;
+    for (ObjectMonitor* monitor: delete_list) {
```

Comment:
I didn't realize C++ has a "foreach" loop construct! Is this in our allowed C++ usage?

Reply:
@fisk - this one is for you... :-)

Reply:
Yeah this is one of the new cool features we can use. I thought it is allowed, because it is neither in the excluded nor undecided list of features in our doc/hotspot-style.md file.

Reply:
https://bugs.openjdk.java.net/browse/JDK-8254733